src/main/java/net/fabricmc/loader/transformer/EnvironmentStripData.java

```diff
@@ -0,0 +1,165 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.transformer;
+
+import net.fabricmc.api.Environment;
+import net.fabricmc.api.EnvironmentInterface;
+import net.fabricmc.api.EnvironmentInterfaces;
+import org.objectweb.asm.*;
+
+import java.util.Collection;
+import java.util.HashSet;
+
+/**
+ * Scans a class for Environment and EnvironmentInterface annotations to figure out what needs to be stripped.
+ */
+public class EnvironmentStripData extends ClassVisitor {
+	private static final String ENVIRONMENT_DESCRIPTOR = Type.getDescriptor(Environment.class);
+	private static final String ENVIRONMENT_INTERFACE_DESCRIPTOR = Type.getDescriptor(EnvironmentInterface.class);
+	private static final String ENVIRONMENT_INTERFACES_DESCRIPTOR = Type.getDescriptor(EnvironmentInterfaces.class);
+
+	private final String envType;
+
+	private boolean stripEntireClass = false;
+	private final Collection<String> stripInterfaces = new HashSet<>();
+	private final Collection<String> stripFields = new HashSet<>();
+	private final Collection<String> stripMethods = new HashSet<>();
+
+	private class EnvironmentAnnotationVisitor extends AnnotationVisitor {
+		private final Runnable onEnvMismatch;
+
+		private EnvironmentAnnotationVisitor(int api, Runnable onEnvMismatch) {
+			super(api);
+			this.onEnvMismatch = onEnvMismatch;
+		}
+
+		@Override
+		public void visitEnum(String name, String descriptor, String value) {
+			if ("value".equals(name) && !envType.equals(value)) {
+				onEnvMismatch.run();
+			}
+		}
+	}
+
+	private class EnvironmentInterfaceAnnotationVisitor extends AnnotationVisitor {
+		private boolean envMismatch;
+		private Type itf;
+
+		private EnvironmentInterfaceAnnotationVisitor(int api) {
+			super(api);
+		}
+
+		@Override
+		public void visitEnum(String name, String descriptor, String value) {
+			if ("value".equals(name) && !envType.equals(value)) {
+				envMismatch = true;
+			}
+		}
+
+		@Override
+		public void visit(String name, Object value) {
+			if ("itf".equals(name)) {
+				itf = (Type) value;
+			}
+		}
+
+		@Override
+		public void visitEnd() {
+			if (envMismatch) {
+				stripInterfaces.add(itf.getInternalName());
+			}
+		}
+	}
+
+	private AnnotationVisitor visitMemberAnnotation(String descriptor, boolean visible, Runnable onEnvMismatch) {
+		if (ENVIRONMENT_DESCRIPTOR.equals(descriptor)) {
+			return new EnvironmentAnnotationVisitor(api, onEnvMismatch);
+		}
+		return null;
+	}
+
+	public EnvironmentStripData(int api, String envType) {
+		super(api);
+		this.envType = envType;
+	}
+
+	@Override
+	public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
+		if (ENVIRONMENT_DESCRIPTOR.equals(descriptor)) {
```

Comment:
Can we use one of those fancy String switch-cases here?

Reply:
String switches requires the cases to be constant, which ENVIRONMENT_DESCRIPTOR and family are unfortunately not.